File: packages/llm-test-harness/src/eval/evaluator.ts
import type { EvalResult, MetricFn, MetricScore } from '../types.js'

export async function runEval(text: string, metrics: MetricFn[]): Promise<EvalResult> {
  if (metrics.length === 0) {
    return { pass: true, passRate: 1, scores: [] }
  }

  const scores: MetricScore[] = await Promise.all(
    metrics.map((m) => Promise.resolve(m.evaluate(text))),
  )

  const passed = scores.filter((s) => s.pass).length
  const passRate = passed / scores.length
  const pass = passRate === 1

  return { pass, passRate, scores }
}

File: packages/llm-test-harness-python/llm_test_harness/eval/evaluator.py
from __future__ import annotations

from ..types import EvalResult, MetricFn, MetricScore


def evaluate(text: str, metrics: list[MetricFn]) -> EvalResult:
    scores: list[MetricScore] = []
    for metric in metrics:
        result = metric.evaluate(text)
        scores.append(result)

    if not scores:
        return EvalResult(passed=True, pass_rate=1.0, scores=[])

    passed_count = sum(1 for s in scores if s.passed)
    pass_rate = passed_count / len(scores)
    return EvalResult(
        passed=passed_count == len(scores),
        pass_rate=pass_rate,
        scores=scores,
    )
